tests/tracer/test_encoders.py

```diff
@@ -299,6 +299,34 @@ def allencodings(f):
     return pytest.mark.parametrize("encoding", MSGPACK_ENCODERS.keys())(f)
 
 
+def test_msgpack_encoding_after_an_exception_was_raised():
+    """Ensure that the encoder's state is consistent after an Exception is raised during encoding"""
+    # Encode a trace after a rollback/BufferFull occurs exception
+    rolledback_encoder = MsgpackEncoderV05(1 << 12, 1 << 12)
+    trace = gen_trace(nspans=1, ntags=100, nmetrics=100, key_size=10, value_size=10)
+    rand_string = rands(size=20, chars=string.ascii_letters)
+    trace[-1].set_tag_str("some_tag", rand_string)
```

Comment:
> don't we want to write this random string into the first span? if the trace is much larger than the buffer we'll bail before we get to the last span, right?

The trace only has one span now. So we are writing the random string in the first span. I will make this more clear.

> Are we sure rand_string is even being written into the string table at all?

Yes this tag is adding the the random string to the StringTable. If we fail to add strings to a string table a ValueError is raised ([link](https://github.com/DataDog/dd-trace-py/blob/4bff470887108a52aeacbc6d79d2c1ec2a501319/ddtrace/internal/_encoding.pyx#L262)). This will crash the test since we only except `BufferItemTooLarge` to be raised.  This test does not overflow the string. It would overflows the trace buffer.

does that matter with this bug?

Yes. If `rand_string` is not added this test will fail.

Reply:
oh, I think I missed the fact that this is one single span....

and... ah... I think the one piece I didn't understand is "when do we bail with buffer too full", it looks like we always encode the full span, and then check _after_ it is encoded if the buffer is now overflowed and then roll back... is that correct?

Reply:
The string table maps strings to indices and encodes this mapping as bytes via msgpack. The "trace buffer" encodes spans as a list of integers and also encodes this data via msgpack. In my reproduction we essentially want the trace buffer to overflow without overflowing the string table.

> I think the one piece I didn't understand is "when do we bail with buffer too full"

We bail on two scenarios:
  1) The string table is full. This [raises](https://github.com/DataDog/dd-trace-py/blob/4bff470887108a52aeacbc6d79d2c1ec2a501319/ddtrace/internal/_encoding.pyx#L262) a ValueError and we immediately stop encoding the trace. We are not testing this scenario.
  2) The trace buffer is full. We only detect this after the trace is already fully encoded ([link](https://github.com/DataDog/dd-trace-py/blob/4bff470887108a52aeacbc6d79d2c1ec2a501319/ddtrace/internal/_encoding.pyx#L506-#L519)). However since this test only encodes one very large span, we actually catch `BufferItemTooLarge`.  But it's essentially the same thing.

Reply:
Another fix we could've implemented is to never rollback and allow payloads to surpass the max size. This was actually my first fix and how I isolated this issue to the rollback functionality. However since the agent can only accept payloads less than 50MB this could be dangerous/inefficient.